### PR TITLE
Simulation fixes, enums, logging, a state machine, node linking, panning, and zooming!

### PR DESCRIPTION
**Pull request by djtran**
When adding nodes, the graph doesn't fly everywhere. The lines are now
being drawn differently, will need to revise to be drawn behind the
nodes. All in all, the graphing javascript has been condensed and is
easier to read.

### CODE REVIEW

**Comment by djtran** [comment]
# Don't merge until link data has been added to this PR. Opened early for comment and review

**Comment by djtran** [comment]
I have yet to test the latest two commits, but functionally the only difference should be you can't create a node on top of another node. All I did was add more comments, logging functionality, and introduced a state machine that will affect behavior (like clicking on another node vs on empty space).

**Comment by djtran** [comment]
Tested, logging and linking work. The ipify stuff isn't however. Debating on just completely ripping it out. thoughts? 

There are a lot of places where this can be improved, both style-wise and function-wise. Don't hesitate to point things out or make suggestions!

File: app/assets/javascripts/GraphManager.js [hunk 1]
```diff
@@ -8,25 +11,44 @@ var completeGraph = {
...
-var nodeRadius = 2;
-var nodeMinDist = 60;
-var nodeMaxDist = nodeMinDist*3;
+var state = STATE.IDLE;
+var selection = {
+    type: TYPE.EMPTY.value
+};
+var zoomTransform;
+
+//For special states (drag or linkstart)
```
Comment (jayajr) [inline app/assets/javascripts/GraphManager.js]: What's the difference between state and special_state? Do we need two?

File: app/assets/javascripts/GraphManager.js [hunk 2]
```diff
@@ -40,99 +62,109 @@ $(document).ready(function(){
...
     setTimeout(function() {
         simulation.force("center", null);
     }, 1000);
 
-    $("svg").on("click", function(event) {
-        addNode(event);
-    });
+    setControls();
+
+    //Put it into a setTimeout cause I'm not sure if the call is blocking, thoughts? If for whatever reason this
```
Comment (jayajr) [inline app/assets/javascripts/GraphManager.js]: What does setTimeout do? If the site does go down, leave it to the user to reenter stuff? DB API is called whenever stuff is created anyways right?

File: app/assets/javascripts/GraphManager.js [hunk 3]
```diff
@@ -142,7 +174,132 @@ function dragged(d) {
...
+    });
+
+    $("svg").on("contextmenu", function(event) {
+        //Disable that default context menu shit
+        event.preventDefault();
+        switch (state) {
+            case STATE.HOVEREMPTY:
+                // TODO: Context Menu?
+                break;
+            case STATE.HOVERNODE:
```
Comment (jayajr) [inline app/assets/javascripts/GraphManager.js]: So special state is just for hovering/linking?

Comment (djtran) [reply app/assets/javascripts/GraphManager.js]: It's more for things like linking, where you have to start the link but then release it on top of another node.

Comment (djtran) [reply app/assets/javascripts/GraphManager.js]: In general, special state is a state that requires a second state or target

File: app/assets/javascripts/GraphManager.js (hunk 2 above)
Comment (djtran) [reply app/assets/javascripts/GraphManager.js]: Set timeout just sets a delayed event, here it says 1 second later release the centering force on the deck simulation

Comment (djtran) [reply app/assets/javascripts/GraphManager.js]: D3 simulation* also I didn't want the application to wait for what I thought was a possibly blocking ajax call, so I put it into setTimeout which is async.

Comment (djtran) [reply app/assets/javascripts/GraphManager.js]: I was also referring to the ipify site, but if our service goes down we should also make that apparent to the user. Something to add to the TODO

**Pull request merged**